examples/step-67/step-67.cc

```diff
@@ -0,0 +1,2305 @@
+  // \max\left(\|\mathbf{u}^-\|+c^-, \|\mathbf{u}^+\|+c^+\right)$ gives the
+  // maximal wave speed and $c = \sqrt{\lambda p / \rho}$ is the speed of
+  // sound. Here, we choose two modifications of that expression for reasons
+  // of computational efficiency, given the small impact of the flux on the
+  // solution. For the above definition of the factor $\lambda$, we would need
+  // to take four square roots, two for the two velocity norms and two for the
+  // speed of sound on either side. The first modification is hence to rather
+  // use $\sqrt{\|\mathbf{u}\|^2+c^2}$ as an estimate of the maximal speed
+  // (which is at most a factor of 2 away from the actual maximum, as shown in
+  // the introduction). This allows us to pull the square root out of the
+  // maximum and get away with a single square root computation. The second
+  // modification is to further relax on the parameter $\lambda$---the smaller
+  // it is, the smaller the dissipation factor (which is multiplied by the
+  // jump in $\mathbf{w}$, which might result in a smaller or bigger
+  // dissipation in the end). This allows us to fit the spectrum into the
+  // stability region of the explicit Runge--Kutta integrator with bigger time
+  // steps. However, we cannot make dissipation too small because otherwise
+  // imaginary eigenvalues grow larger. Finally, the current conservative
+  // formulation is not energy-stable in the limit of $\lambda\to 0$ as it is
+  // not skew-symmetric, and would need additional measures such as split-form
+  // DG schemes in that case.
+  //
+  // For the HLL flux, we follow the formula from literature, introducing an
+  // additional weighting of the two states from Lax--Friedrichs by a
+  // parameter $s$. It is derived from the physical transport directions of
+  // the Euler equations in terms of the current direction of velocity and
+  // sound speed. For the velocity, we here choose a simple arithmetic average
+  // which is sufficient for DG scenarios and moderate jumps in material
+  // parameters.
+  //
+  // Since the numerical flux is multiplied by the normal vector in the weak
+  // form, we multiply by the result by the normal vector for all terms in the
+  // equation. In these multiplications, the `operator*` defined above enables
+  // a compact notation similar to the mathematical definition.
+  template <int dim, typename Number>
+  inline DEAL_II_ALWAYS_INLINE //
+    Tensor<1, dim + 2, Number>
+    euler_numerical_flux(const Tensor<1, dim + 2, Number> &u_m,
+                         const Tensor<1, dim + 2, Number> &u_p,
+                         const Tensor<1, dim, Number> &    normal)
+  {
+    const auto velocity_m = euler_velocity<dim>(u_m);
+    const auto velocity_p = euler_velocity<dim>(u_p);
+
+    const auto pressure_m = euler_pressure<dim>(u_m);
+    const auto pressure_p = euler_pressure<dim>(u_p);
+
+    const auto flux_m = euler_flux<dim>(u_m);
+    const auto flux_p = euler_flux<dim>(u_p);
```

Comment:
It does indeed, at least gcc. Here is the assembly output of the loop over quadrature points for `local_apply_face`, which includes the numerical flux:
```assembly
.L11816:
	movq	-25560(%rbp), %r15
	vmovapd	(%r15), %zmm6 ; normal vector x
	vmovapd	64(%r15), %zmm5 ; normal vector y
	vmovapd	128(%r15), %zmm4 ; normal vector z
.L11772:
	movq	-25464(%rbp), %rsi
	movq	-25496(%rbp), %r12
	movq	-25520(%rbp), %r15
	leaq	0(,%rdx,8), %rax
	movq	-25472(%rbp), %r14
	movq	-25504(%rbp), %r11
	leaq	(%rsi,%rax), %r10
	leaq	(%r12,%rax), %rsi
	leaq	(%r15,%rax), %r12
	vmovapd	(%r12), %zmm13
	vmovapd	.LC70(%rip), %zmm0
	movq	-25480(%rbp), %r13
	movq	-25512(%rbp), %rbx
	leaq	(%r14,%rax), %r9
	leaq	(%r11,%rax), %r14
	vdivpd	(%r14), %zmm0, %zmm22 ; inverse density interior side
	vmulpd	%zmm13, %zmm22, %zmm16
	leaq	0(%r13,%rax), %r8
	leaq	(%rbx,%rax), %r13
	vmovapd	0(%r13), %zmm1
	movq	-25528(%rbp), %rbx
	vmulpd	%zmm1, %zmm22, %zmm10
	vmulpd	%zmm13, %zmm16, %zmm23
	addq	%rax, %rbx
	vmovapd	(%rbx), %zmm12
	movq	-25536(%rbp), %r11
	vmovapd	%zmm10, %zmm17
	vmulpd	%zmm12, %zmm22, %zmm2 ; velocity interior side ready
	vfmadd132pd	%zmm1, %zmm23, %zmm17
	vmovapd	%zmm12, %zmm9
	addq	%rax, %r11
	vmovapd	(%r11), %zmm26
	vmovapd	(%r8), %zmm14
	vfmadd132pd	%zmm2, %zmm17, %zmm9
	vfmadd231pd	%zmm12, %zmm2, %zmm17
	vmulpd	%zmm1, %zmm10, %zmm20 
	vdivpd	(%r10), %zmm0, %zmm21 ; inverse density exterior side
	vmulpd	%zmm14, %zmm21, %zmm15
	vmovapd	(%r9), %zmm3
	vfnmadd132pd	%zmm25, %zmm26, %zmm17
	vmulpd	%zmm3, %zmm21, %zmm8
	movq	-25488(%rbp), %rdi
	vmulpd	%zmm14, %zmm15, %zmm19
	addq	%rax, %rdi
	vfmadd231pd	%zmm24, %zmm17, %zmm20
	vmovapd	(%rdi), %zmm18
	vmulpd	%zmm10, %zmm12, %zmm30
	vfmadd231pd	%zmm24, %zmm17, %zmm23
	vmovapd	%zmm8, %zmm11
	vmovapd	%zmm20, -25008(%rbp)
	vmovapd	%zmm17, %zmm20
	vfmadd132pd	%zmm24, %zmm26, %zmm20
	vmulpd	%zmm18, %zmm21, %zmm7 ; velocity exterior side ready
        ; what comes next should be the evaluation of flux_m and flux_p (already started before)
	vfmadd132pd	%zmm3, %zmm19, %zmm11
	vmovapd	%zmm30, -24944(%rbp)
	vmovapd	%zmm23, %zmm30
	vmulpd	%zmm20, %zmm16, %zmm23
	vmulpd	%zmm20, %zmm10, %zmm28
	vfmadd231pd	%zmm18, %zmm7, %zmm11
	vmulpd	%zmm16, %zmm12, %zmm29
	vmulpd	%zmm13, %zmm10, %zmm31
	vmovapd	%zmm23, -25200(%rbp)
	vmulpd	%zmm2, %zmm12, %zmm23
	vfnmadd213pd	(%rsi), %zmm25, %zmm11
	vmovapd	%zmm28, -25072(%rbp)
	vmovapd	%zmm29, -25136(%rbp)
	vmulpd	%zmm1, %zmm2, %zmm28
	vfmadd132pd	%zmm24, %zmm23, %zmm17
	vmulpd	%zmm24, %zmm11, %zmm0
	vmulpd	%zmm13, %zmm2, %zmm29
	vfnmadd132pd	%zmm25, %zmm26, %zmm9
	vfmadd132pd	%zmm10, %zmm27, %zmm10
	vmovapd	%zmm17, -25392(%rbp)
	vmulpd	%zmm20, %zmm2, %zmm17
	vmulpd	%zmm14, %zmm8, %zmm20
	vmovapd	%zmm0, -24496(%rbp)
	vmulpd	%zmm3, %zmm7, %zmm0
	vmovapd	%zmm31, -24880(%rbp)
	vmovapd	%zmm28, -25264(%rbp)
	vmovapd	%zmm29, -25328(%rbp)
	vmovapd	%zmm17, -25456(%rbp)
	vmovapd	%zmm20, -24560(%rbp)
	vmovapd	%zmm11, %zmm20
	vfmadd213pd	(%rsi), %zmm24, %zmm20
	vmovapd	%zmm0, -24752(%rbp)
	vmulpd	%zmm24, %zmm9, %zmm9
	vmulpd	%zmm14, %zmm7, %zmm0
	vfmadd231pd	%zmm16, %zmm16, %zmm10
	vmulpd	%zmm18, %zmm8, %zmm28
	vmulpd	%zmm20, %zmm8, %zmm29
	vmulpd	%zmm3, %zmm8, %zmm17
	vmulpd	.LC82(%rip), %zmm9, %zmm9
	vmovapd	%zmm0, -24816(%rbp)
	vfmadd132pd	%zmm8, %zmm27, %zmm8
	vmulpd	%zmm18, %zmm7, %zmm0
	vfmadd231pd	%zmm2, %zmm2, %zmm10
	vmovapd	-24496(%rbp), %zmm2
	vfmadd231pd	%zmm24, %zmm11, %zmm17
	vfmadd231pd	%zmm24, %zmm11, %zmm19
	vfmadd231pd	%zmm15, %zmm15, %zmm8
	vfmadd132pd	%zmm24, %zmm0, %zmm11
	vmulpd	%zmm22, %zmm9, %zmm0 ; use inverse density m for lambda
	vmulpd	.LC82(%rip), %zmm2, %zmm9
	vsubpd	%zmm14, %zmm13, %zmm2
	vfmadd231pd	%zmm7, %zmm7, %zmm8
	vmulpd	%zmm14, %zmm5, %zmm14
	vaddpd	%zmm10, %zmm0, %zmm10
	vmulpd	%zmm21, %zmm9, %zmm0  ; use inverse density p for lambda
	vmulpd	%zmm1, %zmm16, %zmm31
	vmulpd	%zmm20, %zmm15, %zmm23
	vfmadd231pd	%zmm3, %zmm6, %zmm14
	vmovapd	%zmm28, -24624(%rbp)
	vaddpd	%zmm8, %zmm0, %zmm8
	vmulpd	%zmm18, %zmm15, %zmm28
	vmovapd	%zmm29, -24688(%rbp)
	vmulpd	%zmm3, %zmm15, %zmm29
	vmaxpd	%zmm10, %zmm8, %zmm15 ; max for lambda
	vmovapd	(%r14), %zmm10
	vfmadd231pd	%zmm18, %zmm4, %zmm14
	vmulpd	%zmm31, %zmm5, %zmm21
	vsubpd	%zmm18, %zmm12, %zmm0
	vmulpd	%zmm20, %zmm7, %zmm20
	vmulpd	%zmm29, %zmm5, %zmm18
	vsqrtpd	%zmm15, %zmm7 ; sqrt for lambda
        ; the following operations involving zmm4, zmm5, zmm6 multiply the fluxes by the normal
	vmulpd	%zmm28, %zmm5, %zmm22
	vsubpd	%zmm3, %zmm1, %zmm15
	vmulpd	%zmm23, %zmm5, %zmm16
	vmulpd	%zmm13, %zmm5, %zmm3
	vmulpd	-25136(%rbp), %zmm5, %zmm31
	vsubpd	(%r10), %zmm10, %zmm8
	vsubpd	(%rsi), %zmm26, %zmm10
	vmulpd	%zmm19, %zmm5, %zmm26
	vmulpd	%zmm30, %zmm5, %zmm19
	vmulpd	-25200(%rbp), %zmm5, %zmm5
	vmulpd	%zmm25, %zmm7, %zmm9
	vfmadd231pd	%zmm17, %zmm6, %zmm18
	vfmadd132pd	%zmm6, %zmm3, %zmm1
	vfmadd231pd	-24560(%rbp), %zmm6, %zmm26
	vfmadd231pd	-24624(%rbp), %zmm6, %zmm22
	vfmadd231pd	-24688(%rbp), %zmm6, %zmm16
	vfmadd231pd	-25008(%rbp), %zmm6, %zmm21
	vfmadd231pd	-24880(%rbp), %zmm6, %zmm19
	vfmadd231pd	-24944(%rbp), %zmm6, %zmm31
	vfmadd132pd	-25072(%rbp), %zmm5, %zmm6
	vmulpd	%zmm25, %zmm9, %zmm9
	vfmadd132pd	%zmm4, %zmm1, %zmm12
	vfmadd132pd	%zmm4, %zmm22, %zmm11
	vfmadd231pd	-24752(%rbp), %zmm4, %zmm18
	vfmadd231pd	-24816(%rbp), %zmm4, %zmm26
	vfmadd132pd	%zmm4, %zmm16, %zmm20
	vfmadd231pd	-25264(%rbp), %zmm4, %zmm21
	vfmadd231pd	-25328(%rbp), %zmm4, %zmm19
	vfmadd231pd	-25392(%rbp), %zmm4, %zmm31
	vfmadd132pd	-25456(%rbp), %zmm6, %zmm4
        ; compute lambda * (u_m - u_p)
	vmulpd	%zmm15, %zmm9, %zmm7
	vmulpd	%zmm8, %zmm9, %zmm8
	vmulpd	%zmm2, %zmm9, %zmm15
	vmulpd	%zmm0, %zmm9, %zmm0
	vmulpd	%zmm10, %zmm9, %zmm9
        ; add flux_m * normal + flux_p * normal
	vaddpd	%zmm14, %zmm12, %zmm6
	vaddpd	%zmm11, %zmm31, %zmm1
	vaddpd	%zmm20, %zmm4, %zmm29
	vaddpd	%zmm18, %zmm21, %zmm30
	vaddpd	%zmm26, %zmm19, %zmm17
        ; add 0.5 * [lambda * (u_m - u_p)] to 0.5 * (flux_m * normal + flux_p * normal)
	vfmadd231pd	%zmm6, %zmm25, %zmm8
	vfmadd231pd	%zmm30, %zmm25, %zmm7
	vfmadd231pd	%zmm17, %zmm25, %zmm15
	vfmadd231pd	%zmm1, %zmm25, %zmm0
	vfmadd231pd	%zmm29, %zmm25, %zmm9
	cmpb	$1, -24420(%rbp)
        ; switch sign for flux on m side
	vsubpd	%zmm8, %zmm27, %zmm13
	vsubpd	%zmm7, %zmm27, %zmm12
	vsubpd	%zmm15, %zmm27, %zmm4
	vsubpd	%zmm0, %zmm27, %zmm2
	vsubpd	%zmm9, %zmm27, %zmm11
        ; the following jumps are for the affine mesh/general mesh case on the two sides
	jbe	.L11814
        ; this is the general mesh case (different JxW in each point) for eval_m
	movq	-25552(%rbp), %r15
	vmovapd	(%r15,%rdx,8), %zmm10
	vmulpd	%zmm13, %zmm10, %zmm14
	vmulpd	%zmm12, %zmm10, %zmm3
	vmulpd	%zmm4, %zmm10, %zmm5
	vmulpd	%zmm2, %zmm10, %zmm6
	vmulpd	%zmm11, %zmm10, %zmm1
	vmovapd	%zmm14, (%r14)
	vmovapd	%zmm3, 0(%r13)
	vmovapd	%zmm5, (%r12)
	vmovapd	%zmm6, (%rbx)
	vmovapd	%zmm1, (%r11)
.L11774:
	cmpb	$1, -23172(%rbp)
	jbe	.L11815
        ; this is the general mesh case (different JxW in each point) for eval_m
	movq	-25544(%rbp), %r14
	vmovapd	(%r14,%rax), %zmm28
	vmulpd	%zmm8, %zmm28, %zmm8
	vmulpd	%zmm7, %zmm28, %zmm7
	vmulpd	%zmm15, %zmm28, %zmm15
	vmulpd	%zmm0, %zmm28, %zmm0
	vmulpd	%zmm9, %zmm28, %zmm9
	vmovapd	%zmm8, (%r10)
	vmovapd	%zmm7, (%r9)
	vmovapd	%zmm15, (%r8)
	vmovapd	%zmm0, (%rdi)
	vmovapd	%zmm9, (%rsi)
.L11776:
	addq	$8, %rdx
	addq	$192, %rcx
	cmpq	%rdx, -24432(%rbp)
	je	.L11777
.L11778:
	cmpb	$2, -24420(%rbp)
	jbe	.L11816
	vmovapd	(%rcx), %zmm6 ; normal vector x for next round
	vmovapd	64(%rcx), %zmm5 ; normal vector y for next round
	vmovapd	128(%rcx), %zmm4 ; normal vector y for next round
	jmp	.L11772
	.p2align 4,,10
	.p2align 3
.L11815:
        ; this multiplies by JxW for the Cartesian/affine mesh case
	movq	-23200(%rbp), %rax
	movq	-25544(%rbp), %r13
	vbroadcastsd	(%rax,%rdx), %zmm3
	vmulpd	0(%r13), %zmm3, %zmm5
	vmulpd	%zmm8, %zmm5, %zmm6
	vmulpd	%zmm7, %zmm5, %zmm1
	vmulpd	%zmm15, %zmm5, %zmm10
	vmulpd	%zmm0, %zmm5, %zmm14
	vmulpd	%zmm9, %zmm5, %zmm13
	vmovapd	%zmm6, (%r10)
	vmovapd	%zmm1, (%r9)
	vmovapd	%zmm10, (%r8)
	vmovapd	%zmm14, (%rdi)
	vmovapd	%zmm13, (%rsi)
	jmp	.L11776
	.p2align 4,,10
	.p2align 3
.L11814:
        ; this multiplies by JxW for the Cartesian/affine mesh case
	movq	-23856(%rbp), %r15
	vbroadcastsd	(%r15,%rdx), %zmm10
	movq	-25552(%rbp), %r15
	vmulpd	(%r15), %zmm10, %zmm14
	vmulpd	%zmm13, %zmm14, %zmm13
	vmulpd	%zmm12, %zmm14, %zmm12
	vmulpd	%zmm4, %zmm14, %zmm4
	vmulpd	%zmm2, %zmm14, %zmm2
	vmulpd	%zmm11, %zmm14, %zmm11
	vmovapd	%zmm13, (%r14)
	vmovapd	%zmm12, 0(%r13)
	vmovapd	%zmm4, (%r12)
	vmovapd	%zmm2, (%rbx)
	vmovapd	%zmm11, (%r11)
	jmp	.L11774  
```
I have checked the formulas involved here (even though reading assembly is tough) and I cannot spot redundant operations here. There are exactly two divisions and one square root (for the modified Lax-Friedrichs flux), no function calls in between the quadrature point region, and jumps only due to the decision in `FEEvaluationBase::submit_value()` to distinguish the plus/minus side.

Reply:
Of course, this kind of assembly code is incredibly hard to read because the compiler schedules operations out of order. But overall, I think the assembly code is super clean.

Reply:
Thanks: I am convinced!